src/modules/tensor/hip/kernel/median_filter.cpp

```diff
@@ -0,0 +1,1906 @@
+/*
+MIT License
+
+Copyright (c) 2019 - 2025 Advanced Micro Devices, Inc.
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "hip_tensor_executors.hpp"
+
+// -------------------- median_filter device helper --------------------
+template <int kernelSize>
+__device__ void median_filter_row_hip_compute(uchar *srcPtr, d_float8 *median_f8)
+{
+    const int paddedKernelWidth = kernelSize + 7; // padded row size for aligned memory access
+    const int uint32LoadCountPerRow = (kernelSize + 10) / 4; // Number of 32-bit loads required to read each row
+    const int windowSize = kernelSize * kernelSize; // Total number of pixels in the filter window
+    const int medianIndex = (windowSize - 1) / 2; // Index of the median value in the sorted list
+    const int sortSteps = medianIndex + 1; // Number of steps required for partial sort to find the median
+
+    float src[kernelSize * paddedKernelWidth], medianVals[8];
+
+    // Load and unpack image data into float using bitwise operations
+    for (int i = 0; i < kernelSize; ++i)
+    {
+        uchar *rowPtr = srcPtr + i * SMEM_LENGTH_X; // Pointer to the start of the current row
+        uint32_t *uintPtr = (uint32_t *)rowPtr;
+
+        // Read the row in 32-bit chunks and unpack 4 uchar values from each
+        for (int j = 0; j < uint32LoadCountPerRow; ++j)
+        {
+            uint32_t val = uintPtr[j];
```

Comment:
uintPtr or srcPtrRowUint[j] is already shared mem and reads should be fast, no need to reallocate into val.